eventuals/task.h

```diff
@@ -101,14 +127,22 @@ struct HeapTask final {
     // 'Register()' more than once is well-defined.
     adapted_.Register(interrupt);
 
-    adapted_.Fail(std::move(exception));
+    std::visit(
+        [this](auto&& error) {
+          if constexpr (!std::is_same_v<
+                            std::decay_t<decltype(error)>,
+                            std::monostate>) {
```

Comment:
As called out in `generator.h`, why do we need both this `if constexpr` and the one in the dispatch callback? Seems like we should only need one!